### DISCUSSION TITLE
API for grasslands for habitat suitability final project

### DISCUSSION THREAD

**Discussion by hanried** [earthlab-education/Earth-Analytics-AY24#639] [General]
In case anyone isn't sure what API to use to get the grassland data, I used this one: https://data-usfs.hub.arcgis.com/datasets/usfs::national-grassland-units-feature-layer/api 

And here's the code I used to use said API:
# Set up the grassland URL
grasslands_url = ("https://apps.fs.usda.gov/arcx/rest/services"
    "/EDW/EDW_NationalGrassland_01/MapServer/0"
    "/query?where=1%3D1&outFields=*&geometry=&geometry"
    "Type=esriGeometryEnvelope&inSR=4326&spatialRel=esriSpatialRelIntersects&outSR=4326&f=json")

# Set up a path to save the data on your machine
grasslands_dir = os.path.join(data_dir, 'grasslands_dir')

# Make the grasslands directory
os.makedirs(grasslands_dir, exist_ok=True)

# Join grasslands shapefile path
grasslands_path = os.path.join(grasslands_dir, 'S_USA.NationalGrassland.shp')

# Only download once
if not os.path.exists(grasslands_path):
    grasslands_gdf = gpd.read_file(grasslands_url)
    grasslands_gdf.to_file(grasslands_path)

# Check that the URL is correct
grasslands_gdf = gpd.read_file(grasslands_path)

# Display grasslands_gdf and plot grasslands to check download
display(grasslands_gdf)
grasslands_gdf.plot()

<img width="452" alt="image" src="https://github.com/user-attachments/assets/cf355f4c-96b3-4bd3-80e9-c20aa872fb20">

I haven't gotten past this step of the project yet but hopefully it's the right API!

**Comment by eculler** [comment]
Cool, @hanried ! This looks great. I also put a direct link in the assignment description if you prefer to download a single shapefile with all the grasslands rather than to go through the API (or if this doesn't end up working out for you)

**Comment by BioNCphilic** [comment]
Thanks! A bit ahead of where I am, but nice to have a guidepost.

**Comment by BioNCphilic** [comment]
@eculler - good to know it is ok to download a single shapefile. I'm doing this a bit differently because I want to compare two areas which can support native rivercane which both the Eastern band of the Cherokee & the Cherokee Nation use for a variety of purposes. It's actually pretty cool that rivercane can grow as far west as Oklahoma. However, when I looked at the grassland data the area in Oklahoma was not anywhere near the area where Cherokee people live. 

So - I hope it's ok if I got the shape files for Cherokee County OK and Jackson County NC from a US Census Source. I downloaded the zip files and was going to use python code to extract the .shp files, but it became a mess with different directories for data and another for the project, so I just used my file manager to extract them and moved them into the correct directory. I'm sure I could have made it work, but this was much easier.

**Reply by BioNCphilic** [reply]
Well - actually - extracting files manually did not work apparently even though I can see the file on github and in my terminal.
![Screenshot 2024-12-06 122832](https://github.com/user-attachments/assets/7221df24-2951-4c66-b49c-aef675c0414c)

Now what? I guess I have to get VScode/python to do the extraction?

**Reply by BioNCphilic** [reply]
![image](https://github.com/user-attachments/assets/0ed7662a-df90-45ba-a4ca-660ac061b82c)

**Comment by BioNCphilic** [comment]
OK - after many tries, I finally have a shapefile for US Counties from the US Census. I'm doing this because I want to compare a county in Oklahoma with one in North Carolina as my two locations. Here's the code if you want to use it:

# Define download URL and file paths
base_url = "https://www2.census.gov/geo/tiger/TIGER2024/COUNTY/tl_2024_us_county.zip"
 
# Create the 'counties' directory if it doesn't exist
os.makedirs(county_dir, exist_ok=True)
county_dir = os.path.join(data_dir, "counties") 

# Download county data (if not already downloaded)
if not os.path.exists(os.path.join(county_dir, "tl_2024_us_county.zip")):
  county_response = requests.get(base_url)
  county_response.raise_for_status()  # Raise exception on download errors

  # Save the downloaded data (compressed) to the specified path
  with open(os.path.join(county_dir, "tl_2024_us_county.zip"), "wb") as county_file:
    county_file.write(county_response.content)

# Extract the downloaded shapefile
with zipfile.ZipFile(os.path.join(county_dir, "tl_2024_us_county.zip"), "r") as county_zip:
  county_zip.extractall(county_dir)

print("Shapefile extracted successfully!")

**Reply by BioNCphilic** [reply]
Actually - use this code! And [here is the link](https://transition.fcc.gov/oet/info/maps/census/fips/fips.txt) to figure out locations - specifically a county in a US State.

# Define the download path (replace with your desired location)
download_path = "\GitHubr\ivercane_data"

# Define the URL of the shapefile
url = "https://www2.census.gov/geo/tiger/TIGER2023/COUNTY/tl_2023_us_county.zip"

# Download the shapefile
response = requests.get(url)
response.raise_for_status()

with open("tl_2023_us_county.zip", "wb") as f:
    f.write(response.content)
# Extract the shapefile
with zipfile.ZipFile("tl_2023_us_county.zip", "r") as zip_ref:
    zip_ref.extractall()

# Load the shapefile into a GeoDataFrame
county_data = gpd.read_file("tl_2023_us_county.shp")

# Filter for a specific counties 
cherokee_county = county_data[(county_data['STATEFP'] == '40') & (county_data['COUNTYFP'] == '021')]
jackson_county = county_data[(county_data['STATEFP'] == '37') & (county_data['COUNTYFP'] == '099')]

# Plot the shapefile
cherokee_county.plot()
